src/devices/src/virtio/queue.rs

```diff
@@ -393,6 +427,82 @@ impl Queue {
         let addr = self.avail_ring.unchecked_add(2);
         Wrapping(mem.read_obj::<u16>(addr).unwrap())
     }
+
+    /// Get the value of the used event field of the avail ring.
+    #[inline(always)]
+    pub fn used_event(&self, mem: &GuestMemoryMmap) -> Wrapping<u16> {
+        // We need to find the `used_event` field from the avail ring.
+        let used_event_addr = self
+            .avail_ring
+            .unchecked_add(u64::from(4 + 2 * self.actual_size()));
+
+        Wrapping(mem.read_obj::<u16>(used_event_addr).unwrap())
+    }
+
+    /// Helper method that writes `val` to the `avail_event` field of the used ring.
+    fn set_avail_event(&mut self, val: u16, mem: &GuestMemoryMmap) {
+        let avail_event_addr = self
+            .used_ring
+            .unchecked_add(u64::from(4 + 8 * self.actual_size()));
+
+        mem.write_obj(val, avail_event_addr).unwrap();
+    }
+
+    /// Try to enable notification events from the guest driver. Returns true if notifications were
+    /// successfully enabled. Otherwise it means that one or more descriptors can still be consumed
+    /// from the available ring and we can't guarantee that there will be a notification. In this
+    /// case the caller might want to consume the mentioned descriptors and call this method again.
+    pub fn try_enable_notification(&mut self, mem: &GuestMemoryMmap) -> bool {
+        // If the device doesn't use notification suppression, we'll continue to get notifications
+        // no matter what.
+        if !self.uses_notif_suppression {
+            return true;
```

Comment:
Why are we double checking this ? In pop_or_enable_notification() we are already doing this.

Reply:
Both methods are public so they can be used independently, and they need to treat this case differently.